pyodide-build/pyodide_build/_f2c_fixes.py

```diff
@@ -17,157 +20,219 @@ def fix_f2c_clapack_calls(f2c_output_name: str):
+        'hpcon', 'hpev', 'hpevd', 'hpevx', 'hpgst', 'hpgv', 'hpgvd', 'hpgvx',
+        'hprfs', 'hpsv', 'hpsvx', 'hptrd', 'hptrf', 'hptri', 'hptrs', 'lacp2',
+        'lahef', 'laqhb', 'laqhe', 'laqhp', 'ptrfs', 'pttrs', 'ungbr', 'ungtr',
+        'unm2l', 'unm2r', 'unmbr', 'unmhr', 'unml2', 'unmlq', 'unmql', 'unmqr',
+        'unmr2', 'unmr3', 'unmrq', 'unmrz', 'unmtr', 'upgtr', 'upmtr',
+    ]
+    lapack_other_names = ['lsame_', 'ilaenv_']
     # fmt: on
-    code = None
-    with open(f2c_output_name, "r") as f:
-        code = f.read()
-        for cur_name in lapack_names:
-            code = re.sub(rf"\b{cur_name}\b", "w" + cur_name, code)
-    if code:
-        with open(f2c_output_name, "w") as f:
-            f.write(code)
+    lapack_names = lapack_other_names
+    for name in lapack_czds_names:
+        lapack_names.extend(f"{l}{name}_" for l in ["c", "z", "d", "s"])
+    for name in lapack_cz_names:
+        lapack_names.extend(f"{l}{name}_" for l in ["c", "z"])
+    for name in lapack_ds_names:
+        lapack_names.extend(f"{l}{name}_" for l in ["d", "s"])
+
+    f2c_output = Path(f2c_output_path)
+    if f2c_output.name == "dfft.c":
+        # dfft.c has a bunch of implicit cast args coming from functions copied
+        # out of future lapack versions. fix_inconsistent_decls will fix all
+        # except string to int.
+        subprocess.check_call(
+            [
+                "patch",
+                str(f2c_output_path),
+                f"../../patches/fix-implicit-cast-args-from-newer-lapack.patch",
+            ]
+        )
+
+    with open(f2c_output, "r") as f:
+        lines = f.readlines()
+    if "id_dist" in f2c_output_path:
+        # Fix implicit casts in id_dist.
+        lines = fix_inconsistent_decls(lines)
+    if "odepack" in f2c_output_path or f2c_output.name == "mvndst.c":
+        # Mark all but one declaration of each struct as extern.
+        if f2c_output.name == "blkdta000.c":
+            # extern marking in blkdata000.c doesn't work properly so we let it
+            # define the one copy of the structs. It doesn't talk about lsa001
+            # at all though, so we need to add a definition of it.
+            lines.append(
+                """
+                struct {    doublereal rownd2, pdest, pdlast, ratio, cm1[12], cm2[5], pdnorm;
+                    integer iownd2[3], icount, irflag, jtyp, mused, mxordn, mxords;
+                } lsa001_;
+                """
+            )
+        else:
+            add_externs_to_structs(lines)
+
+    if f2c_output.name in [
+        "wrap_dummy_g77_abi.c",
+        "_lapack_subroutine_wrappers.c",
+        "_blas_subroutine_wrappers.c",
+    ]:
+        lines = remove_ftnlen_args(lines)
+
+    code = "".join(lines)
+    for cur_name in lapack_names:
+        code = re.sub(rf"\b{cur_name}\b", "w" + cur_name, code)
+
+    with open(f2c_output, "w") as f:
+        f.write(code)
+
+
+def remove_ftnlen_args(lines: List[str]) -> List[str]:
```

Comment:
Any chance you could add doctests for the following functions, which would serve as examples of what is expected to happen?